gossip/privdata/dataretriever.go

```diff
@@ -55,14 +55,25 @@ func (dr *dataRetriever) CollectionRWSet(digests []*protosgossip.PvtDataDigest,
 		// if there is an error getting info from the ledger, we need to try to read from transient store
 		return nil, false, errors.Wrap(err, "wasn't able to read ledger height")
 	}
-	if height <= blockNum {
+
+	// The condition may be true for either commit or reconciliation case when another peer sends a request to retrieve private data.
+	// For the commit case, get the private data from the transient store because the block has not been committed.
+	// For the reconciliation case, it means this peer is further behind the ledger height than the peer
+	// that requested for the private data. Therefore, neither transient store nor ledger has the requested private data.
+	if height <= blockNum { // Check whenever current ledger height is equal or below block sequence num.
 		dr.logger.Debug("Current ledger height ", height, "is below requested block sequence number",
 			blockNum, "retrieving private data from transient store")
-	}
 
-	if height <= blockNum { // Check whenever current ledger height is equal or below block sequence num.
 		results := make(Dig2PvtRWSetWithConfig)
 		for _, dig := range digests {
+			// skip retrieving from transient store for reconciliation requests
+			if isReconciliationRequest(dig) {
```

Comment:
Though this is correct that txID is not included in the case of reconciliation at the caller side, but drawing this conclusion at this function which the the sender side, breaks the principal of isolation. I find previous patch more suitable here that if caller does not supply txID (for whatever reason), skip looking in the transient store. We can certainly write a comment when this can happen but the caller code is free to change in general. For instance, caller is free not to send txID even in the case of commit path. In fact, commit path/ reconciliation terms are meaningful only at the caller side. This place is just getting asked for data either by txID, or by blockNum:TxNum (or both).

Also, having this if condition in a loop, may give someone a wrong impression that the requests for reconciliation and commit path gets combined on the caller side.

@ale-linux - second thought?

Reply:
I trust your judgement since you certainly understand the pvtdata code better than I do. I just don't like these less-than-transparent conditions where the behaviour of a function is determined by special values of a field that otherwise plays a completely different role in the system (case in point, the txid's principal meaning is to avoid replay attacks and provide a unique handle to a transaction, and it is here used to determine whether data should be retrieved from storage).

These non-obvious relationships (here between txis(=uniquness) and behaviour of pvt data retrieval) tend to be quickly forgotten and make the code hard to maintain.

That said, if you believe this to be the right approach, I won't stand in the way.